File: loop_optimizations/loop_unrolling.cpp
// Loop Unrolling Example
// Demonstrates manual unrolling of a data-driven loop over an array of
// unknown length.  The key challenge is handling the "tail" — the leftover
// elements when n is not a multiple of the unroll factor.
//
// Three strategies are shown:
//   1. Scalar baseline
//   2. 4x unroll with a scalar cleanup loop for the tail
//   3. Duff's Device — a switch/while hybrid that jumps into the middle of
//      the unrolled body to handle the tail without a separate loop
//
// Build:
//   g++ -O1 -o unroll_O1 loop_unrolling.cpp && ./unroll_O1
//   g++ -O2 -o unroll_O2 loop_unrolling.cpp && ./unroll_O2
//
// -O1 enables register allocation and basic optimisations but not
// auto-vectorisation, so the speedup from manual unrolling is visible.
// -O2 enables auto-vectorisation; the compiler may match or beat the
// hand-unrolled code.

#include <iostream>
#include <chrono>
#include <climits>
#include <iomanip>
#include <string>
#include <vector>

using namespace std;
using namespace std::chrono;

static const int N       = 100'000'000;
static const int RUNS    = 5;

// Prevent the compiler from optimizing away a result.
template <typename T>
static void sink(T const& val) { asm volatile("" : : "m"(val) : "memory"); }

template <typename Func>
long long bench(Func f) {
    long long best = LLONG_MAX;
    for (int r = 0; r < RUNS; r++) {
        asm volatile("" ::: "memory");
        auto t0 = high_resolution_clock::now();
        auto v  = f();
        sink(v);
        auto t1 = high_resolution_clock::now();
        best = min(best, duration_cast<milliseconds>(t1 - t0).count());
    }
    return best;
}

void report(const char* title,
            const char* a_label, long long a_ms,
            const char* b_label, long long b_ms) {
    double speedup = (double)a_ms / max(1LL, b_ms);
    cout << "\n" << string(64, '-') << "\n";
    cout << title << "\n";
    cout << string(64, '-') << "\n";
    cout << "  BASELINE  " << left  << setw(40) << a_label
         << right << setw(5) << a_ms << " ms\n";
    cout << "  UNROLLED  " << left  << setw(40) << b_label
         << right << setw(5) << b_ms  << " ms\n";
    cout << "  Speedup:  " << fixed << setprecision(2) << speedup << "x\n";
}

// ================================================================
// 1. Scalar baseline — one add per iteration
// ================================================================

long long sum_scalar(const int* data, int n) {
    long long s = 0;
    for (int i = 0; i < n; i++)
        s += data[i];
    return s;
}

// ================================================================
// 2. 4x unroll with a scalar cleanup loop
//
// Main loop processes 4 elements at once, reducing loop-control
// overhead (branch, increment) by 4x.  A trailing scalar loop
// handles the remaining 0–3 elements.
// ================================================================

long long sum_unrolled4(const int* data, int n) {
    long long s = 0;
    int i = 0;

    // Process 4 elements per iteration.
    int n4 = n - (n % 4);   // largest multiple of 4 <= n
    for (; i < n4; i += 4) {
        s += data[i];
        s += data[i + 1];
        s += data[i + 2];
        s += data[i + 3];
    }

    // Cleanup: handle tail elements (0, 1, 2, or 3 remaining).
    for (; i < n; i++)
        s += data[i];

    return s;
}

// ================================================================
// 3. Duff's Device — unrolled body with a switch-driven entry point
//
// The switch jumps into the *middle* of the unrolled loop body so
// that the first "iteration" processes exactly (n % 4) elements,
// aligning subsequent iterations to a 4-element boundary.
// No separate cleanup loop is required.
//
// Named after Tom Duff (Bell Labs, 1983) who used it to accelerate
// a byte-copy routine on a PDP-11.
// ================================================================

long long sum_duffs(const int* data, int n) {
    if (n == 0) return 0;

    long long s = 0;
    int i = 0;
    int count = (n + 3) / 4;   // number of times the unrolled body fires

    // switch selects the fall-through entry point for the first pass.
    switch (n % 4) {
        case 0: do { s += data[i++];  // falls through
        case 3:      s += data[i++];  // falls through
        case 2:      s += data[i++];  // falls through
        case 1:      s += data[i++];
                } while (--count > 0);
    }

    return s;
}

// ================================================================
// main — build the array and run benchmarks
// ================================================================

int main() {
    // Use a non-trivial array so the compiler cannot constant-fold results.
    vector<int> data(N);
    for (int i = 0; i < N; i++)
        data[i] = (i % 7) + 1;   // values 1–7, non-power-of-2 length pattern

    // Test with lengths that are *not* multiples of 4 to exercise the tail.
    // n = N - 3  =>  remainder 1
    // n = N      =>  remainder 0  (no tail work needed)
    const int* p = data.data();
    int n_odd = N - 3;   // exercises tail in both unrolled and Duff's versions

    cout << "Array length (odd test): " << n_odd
         << "  (tail = " << (n_odd % 4) << " element(s))\n";
    cout << "Array length (even test): " << N
         << "  (tail = " << (N % 4) << " element(s))\n";

    // Verify all three produce the same answer.
    long long r1 = sum_scalar  (p, n_odd);
    long long r2 = sum_unrolled4(p, n_odd);
    long long r3 = sum_duffs   (p, n_odd);
    cout << "\nCorrectness check (n_odd):\n";
    cout << "  scalar=" << r1 << "  unrolled4=" << r2 << "  duffs=" << r3;
    cout << (r1 == r2 && r2 == r3 ? "  PASS" : "  FAIL") << "\n";

    // ---- Benchmark: odd-length array --------------------------------
    auto t_scalar   = bench([&]{ return sum_scalar  (p, n_odd); });
    auto t_unrolled = bench([&]{ return sum_unrolled4(p, n_odd); });
    auto t_duffs    = bench([&]{ return sum_duffs   (p, n_odd); });

    report("Odd-length array (n % 4 == 1) — loop-overhead reduction",
           "scalar loop",        t_scalar,
           "4x unroll + cleanup", t_unrolled);
    report("Odd-length array (n % 4 == 1) — Duff's Device vs scalar",
           "scalar loop",        t_scalar,
           "Duff's Device",      t_duffs);

    // ---- Benchmark: even-length array (no tail) ---------------------
    auto t_scalar_e   = bench([&]{ return sum_scalar  (p, N); });
    auto t_unrolled_e = bench([&]{ return sum_unrolled4(p, N); });
    auto t_duffs_e    = bench([&]{ return sum_duffs   (p, N); });

    report("Even-length array (n % 4 == 0) — loop-overhead reduction",
           "scalar loop",        t_scalar_e,
           "4x unroll + cleanup", t_unrolled_e);
    report("Even-length array (n % 4 == 0) — Duff's Device vs scalar",
           "scalar loop",        t_scalar_e,
           "Duff's Device",      t_duffs_e);

    cout << "\nNote: at -O2 the compiler auto-vectorises the scalar loop,\n"
            "      often matching or beating manual unrolling.\n"
            "      Build with -O0 to see raw unrolling benefits.\n";
    return 0;
}

File: loop_optimizations/loop_unrolling.rs
// Loop Unrolling Example — Rust port of loop_unrolling.cpp
//
// Three strategies for handling the tail of an array whose length
// is not a multiple of the unroll factor:
//   1. Scalar baseline
//   2. 4x unroll with a scalar cleanup loop
//   3. Duff's Device equivalent (see note below)
//
// Translation notes:
//   - const int* data, int n  →  &[i32]  (slice carries pointer + length)
//   - long long               →  i64
//   - asm volatile sink       →  std::hint::black_box
//   - LLONG_MAX               →  i64::MAX / u128::MAX
//   - vector<int>             →  Vec<i32>
//   - high_resolution_clock   →  std::time::Instant  (returns u128 ms)
//
// Duff's Device cannot be translated mechanically: Rust's match has no
// fallthrough, so a switch that jumps into the middle of a do/while loop
// is inexpressible.  The equivalent behaviour — process the tail first,
// then the aligned bulk — is semantically identical for a commutative
// operation like addition.
//
// Build:
//   rustc -C opt-level=1 -o unroll_O1 loop_unrolling.rs && ./unroll_O1
//   rustc -C opt-level=2 -o unroll_O2 loop_unrolling.rs && ./unroll_O2

use std::hint::black_box;
use std::time::Instant;

const N: usize = 100_000_000;
const RUNS: usize = 5;

// ----------------------------------------------------------------
// Benchmarking helpers
// ----------------------------------------------------------------

fn bench<F: Fn() -> i64>(f: F) -> u128 {
    let mut best = u128::MAX;
    for _ in 0..RUNS {
        let t0 = Instant::now();
        let v = f();
        let elapsed = t0.elapsed().as_millis();
        let _ = black_box(v);  // prevent dead-code elimination
        best = best.min(elapsed);
    }
    best
}

fn report(title: &str, a_label: &str, a_ms: u128, b_label: &str, b_ms: u128) {
    let speedup = a_ms as f64 / b_ms.max(1) as f64;
    println!("\n{}", "-".repeat(64));
    println!("{}", title);
    println!("{}", "-".repeat(64));
    println!("  BASELINE  {:<40} {:>5} ms", a_label, a_ms);
    println!("  UNROLLED  {:<40} {:>5} ms", b_label, b_ms);
    println!("  Speedup:  {:.2}x", speedup);
}

// ================================================================
// 1. Scalar baseline — one add per iteration
// ================================================================

fn sum_scalar(data: &[i32]) -> i64 {
    let mut s: i64 = 0;
    for &x in data {
        s += x as i64;
    }
    s
}

// ================================================================
// 2. 4x unroll with a scalar cleanup loop
//
// Main loop processes 4 elements at once, reducing loop-control
// overhead (branch, increment) by 4x.  A trailing loop handles
// the remaining 0–3 elements.
// ================================================================

fn sum_unrolled4(data: &[i32]) -> i64 {
    let n = data.len();
    let mut s: i64 = 0;
    let n4 = n - (n % 4);   // largest multiple of 4 <= n
    let mut i = 0;

    while i < n4 {
        s += data[i]     as i64;
        s += data[i + 1] as i64;
        s += data[i + 2] as i64;
        s += data[i + 3] as i64;
        i += 4;
    }

    // Cleanup: handle tail elements (0, 1, 2, or 3 remaining).
    while i < n {
        s += data[i] as i64;
        i += 1;
    }

    s
}

// ================================================================
// 3. Duff's Device equivalent
//
// The original C uses switch fallthrough to jump into the middle of
// a do/while loop.  Rust's match does not fall through, so the same
// effect is achieved by:
//   a) processing the tail elements first  (the "entry point" of
//      the C switch), then
//   b) processing the remaining bulk in groups of 4.
//
// The result is identical for addition (commutative, order-independent).
// ================================================================

fn sum_duffs(data: &[i32]) -> i64 {
    let n = data.len();
    if n == 0 { return 0; }

    let mut s: i64 = 0;
    let tail = n % 4;

    // Handle tail first (maps to the switch entry point in the C version).
    for i in 0..tail {
        s += data[i] as i64;
    }

    // Process remaining elements in groups of 4.
    let mut i = tail;
    while i < n {
        s += data[i]     as i64;
        s += data[i + 1] as i64;
        s += data[i + 2] as i64;
        s += data[i + 3] as i64;
        i += 4;
    }

    s
}

// ================================================================
// main — build the array and run benchmarks
// ================================================================

fn main() {
    // Non-trivial values so the compiler cannot constant-fold.
    let data: Vec<i32> = (0..N).map(|i| (i % 7) as i32 + 1).collect();

    let n_odd = N - 3;   // tail = 1 element, exercises cleanup path
    println!("Array length (odd test):  {}  (tail = {} element(s))", n_odd, n_odd % 4);
    println!("Array length (even test): {}  (tail = {} element(s))", N, N % 4);

    // Correctness check.
    let r1 = sum_scalar   (&data[..n_odd]);
    let r2 = sum_unrolled4(&data[..n_odd]);
    let r3 = sum_duffs    (&data[..n_odd]);
    println!("\nCorrectness check (n_odd):");
    print!  ("  scalar={}  unrolled4={}  duffs={}", r1, r2, r3);
    println!("  {}", if r1 == r2 && r2 == r3 { "PASS" } else { "FAIL" });

    // ---- Benchmark: odd-length array --------------------------------
    let t_scalar   = bench(|| sum_scalar   (&data[..n_odd]));
    let t_unrolled = bench(|| sum_unrolled4(&data[..n_odd]));
    let t_duffs    = bench(|| sum_duffs    (&data[..n_odd]));

    report("Odd-length array (n % 4 == 1) — loop-overhead reduction",
           "scalar loop",         t_scalar,
           "4x unroll + cleanup", t_unrolled);
    report("Odd-length array (n % 4 == 1) — Duff's Device equiv. vs scalar",
           "scalar loop",         t_scalar,
           "Duff's Device equiv.", t_duffs);

    // ---- Benchmark: even-length array (no tail) ---------------------
    let t_scalar_e   = bench(|| sum_scalar   (&data));
    let t_unrolled_e = bench(|| sum_unrolled4(&data));
    let t_duffs_e    = bench(|| sum_duffs    (&data));

    report("Even-length array (n % 4 == 0) — loop-overhead reduction",
           "scalar loop",         t_scalar_e,
           "4x unroll + cleanup", t_unrolled_e);
    report("Even-length array (n % 4 == 0) — Duff's Device equiv. vs scalar",
           "scalar loop",         t_scalar_e,
           "Duff's Device equiv.", t_duffs_e);

    println!("\nNote: at opt-level=2 the compiler auto-vectorises the scalar loop,");
    println!("      often matching or beating manual unrolling.");
    println!("      Build with opt-level=1 to see raw unrolling benefits.");
}
